mmdet/models/dense_heads/solov2_head.py

```diff
@@ -0,0 +1,684 @@
+
+    @force_fp32(
+        apply_to=('mlvl_kernel_preds', 'mlvl_cls_scores', 'mask_feats'))
+    def get_results(self, mlvl_kernel_preds, mlvl_cls_scores, mask_feats,
+                    img_metas, **kwargs):
+        """Get multi-image mask results.
+
+        Args:
+            mlvl_kernel_preds (list[Tensor]): Multi-level dynamic kernel
+                prediction. The kernel is used to generate instance
+                segmentation masks by dynamic convolution. Each element in the
+                list has shape
+                (batch_size, kernel_out_channels, num_grids, num_grids).
+            mlvl_cls_scores (list[Tensor]): Multi-level scores. Each element
+                in the list has shape
+                (batch_size, num_classes, num_grids, num_grids).
+            mask_feats (Tensor): Unified mask feature map used to generate
+                instance segmentation masks by dynamic convolution. Has shape
+                (batch_size, mask_out_channels, h, w).
+            img_metas (list[dict]): Meta information of all images.
+
+        Returns:
+            list[:obj:`InstanceData`]: Processed results of multiple
+            images.Each :obj:`InstanceData` usually contains
+            following keys.
+
+                - scores (Tensor): Classification scores, has shape
+                  (num_instance,).
+                - labels (Tensor): Has shape (num_instances,).
+                - masks (Tensor): Processed mask results, has
+                  shape (num_instances, h, w).
+        """
+        num_levels = len(mlvl_cls_scores)
+        assert len(mlvl_kernel_preds) == len(mlvl_cls_scores)
+
+        for lvl in range(num_levels):
+            cls_scores = mlvl_cls_scores[lvl]
+            cls_scores = cls_scores.sigmoid()
+            local_max = F.max_pool2d(cls_scores, 2, stride=1, padding=1)
+            keep_mask = local_max[:, :, :-1, :-1] == cls_scores
+            cls_scores = cls_scores * keep_mask
+            mlvl_cls_scores[lvl] = cls_scores.permute(0, 2, 3, 1)
+
+        result_list = []
+        for img_id in range(len(img_metas)):
+            img_cls_pred = [
+                mlvl_cls_scores[lvl][img_id].view(-1, self.cls_out_channels)
+                for lvl in range(num_levels)
+            ]
+            img_mask_feats = mask_feats[[img_id]]
+            img_kernel_pred = [
+                mlvl_kernel_preds[lvl][img_id].permute(1, 2, 0).view(
+                    -1, self.kernel_out_channels) for lvl in range(num_levels)
+            ]
+            img_cls_pred = torch.cat(img_cls_pred, dim=0)
+            img_kernel_pred = torch.cat(img_kernel_pred, dim=0)
+            result = self._get_results_single(
+                img_kernel_pred,
+                img_cls_pred,
+                img_mask_feats,
+                img_meta=img_metas[img_id])
+            result_list.append(result)
+        return result_list
+
+    def _get_results_single(self,
```

Comment:
Is it possible to reuse the code in SOLOHead?

Reply:
I actually tried but failed.

SOLOv2 uses dynamic convolution to decode the mask, and dynamic convolution cannot be done batch-wise, which means that we have to insert the code of mask decoding into a "do per-image" loop. Function `_get_result_single` is then a perfect candidate.

What's more, SOLO v1 does the score filtering inside the `_get_result_single`. if we reuse the code we would have nowhere to get the filtered indices, and thus we could not rule out the low-score dynamic convolution kernels before dynamic convolution. This would greatly harm the inference speed.

Reply:
There are some differences between v1 and v2, it is hard to reuse SOLOHead.